File: Usine.cs
namespace exercice_trombone
{
    /// <summary>
    /// Usine de fabrication de trombones.
    /// </summary>
    internal class Usine
    {
        /// <summary>
        /// Niveau de qualité des trombones produits.
        /// </summary>
        public int Qualite { get; set; }

        /// <summary>
        /// Quantité de matière consommée pour la production d'un trombone.
        /// </summary>
        public int QuantiteConsommee { get; set;}

        /// <summary>
        /// Vitesse de production des trombones.
        /// </summary>
        public int VitesseProduction { get; set;}

        /// <summary>
        /// Joueur qui possède l'usine.
        /// </summary>
        private Joueur proprietaire;

        /// <summary>
        /// Crée une nouvelle usine de base.
        /// </summary>
        /// <param name="proprietaire">Le joueur qui possède l'usine.</param>
        public Usine(Joueur proprietaire)
        {
            this.proprietaire = proprietaire;
            Qualite = 0;
            QuantiteConsommee = 100;
            VitesseProduction = 1;
        }

        /// <summary>
        /// Achète une amélioration et si possible applique ces effets.
        /// </summary>
        /// <param name="amelioration">L'amélioration à acheter.</param>
        public void AcheterAmelioration(Amelioration amelioration)
        {
            if(proprietaire.Payer(amelioration.Cout))
            {
                // Attendre amelioration.TempsRecherche

                amelioration.AppliquerAmelioration(this);
            }
        }
    }
}

File: src/edu/cegepvicto/usine/Usine.java
package edu.cegepvicto.usine;

import java.util.ArrayList;

/**
 * Usine qui permet la production des trombones.
 */
public class Usine {

    /**
     * Les caractéristiques de l'usine
     */
    private ArrayList<Caracteristique> caracteristiques;

    /**
     * Joueur qui possède l'usine de fabrication des trombones.
     */
    private Joueur proprietaire;

    /**
     * Crée une nouvelle usine de trombones.
     * @param proprietaire le propriétaire de l'usine.
     */
    public Usine(Joueur proprietaire) {
        this.proprietaire = proprietaire;
        caracteristiques = new ArrayList<>();

        caracteristiques.add(new Caracteristique("Qualité", 0));
        caracteristiques.add(new Caracteristique("VitesseProduction", 1));
        caracteristiques.add(new Caracteristique("MatiereConsommee", 100));
     }

    /**
     * Modifie la valeur d'une caractéristique.
     * @param nom le nom de la caractéristique modifiée.
     * @param changement la magnitude du changement de la caractéristique.
     */
    public void changerCaracteristique(String nom, int changement) {
        for(Caracteristique caracteristique : caracteristiques) {
            if(caracteristique.getNom().equals(nom)) {
                caracteristique.setValeur(caracteristique.getValeur() + changement);
            }
        }
    }

    /**
     * Achète l'amélioration et l'applique sur l'usine.
     * @param amelioration l'amélioration achetée.
     */
    public void acheterAmelioration(Amelioration amelioration) {
        if(proprietaire.payer(amelioration.getCout())) {
            // Attendre amelioration.tempsRecherche

            amelioration.appliquerAmelioration(this);
        }
    }

}